internal/cluster/cluster.go

```diff
@@ -207,7 +207,22 @@ func (m *Manager) DeregisterCluster(ctx context.Context, c *common.Cluster) erro
 		return metricsErr
 	}
 
-	err := m.registrationAPI.DeregisterCluster(ctx, c)
+	// Delete discovery image for deregistered cluster
+	discoveryImage := fmt.Sprintf(s3wrapper.DiscoveryImageTemplate, c.ID.String())
+	exists, err := m.objectHandler.DoesObjectExist(ctx, discoveryImage)
+	if err != nil {
+		m.log.WithError(err).Errorf("Failed to find cluster discovery image %s in s3", discoveryImage)
+		return err
+	}
+	if exists {
+		_, err = m.objectHandler.DeleteObject(ctx, discoveryImage)
+		if err != nil {
+			m.log.WithError(err).Errorf("Failed to delete cluster discovery image %s from s3", discoveryImage)
+			return err
```

Comment:
The cluster won't be deregistered if the image deletion fails. Is there a place in the cluster's model/status where we can indicate the reason why it was not deleted?

I think it would be interesting for the operator to know why the cluster was not deregistered as this may be a system-wide issue that would prevent the removal of all the clusters that should be cleaned up (and their images).

Reply:
There are 2 places in which a cluster is being deregistered:
1. By user's request - in which the user gets an immediate feedback.
2. By housekeeping service to clear deregistered clusters.

For the operator, a failure to deregister a cluster is capture by: https://github.com/openshift/assisted-service/blob/master/internal/controller/controllers/clusterdeployments_controller.go#L645
which logs the error and requeue it.
You're raising a valid point regarding a wide implication of deregister failure. In case a special treatment should be added for the operator mode other than requeue, e.g. start collecting metrics for failed attempts to deregister a cluster - that can be added, but needs to be discussed and tracked in a broader context.

Reply:
@masayag I think this is important and we should discuss it. Could you please create an issue where we can start collecting thoughts?

Whether the deregister happens through 1 or 2, the failure to delete the object needs to be properly communicated to the operator of the cloud (may be worth adding a metric?). Otherwise, it may go unnoticed longer than desired taking the deployment to a point where there's no more space available.

Could you also add a TODO/comment in this section with these thoughts (and a link pointing to this discussion in GH? https://github.com/openshift/assisted-service/pull/1443/files#r611667377 <- this one